File: app/src/main/java/bin/mg/main/ui/editor/LinedEditText.java
package bin.mg.main.ui.editor;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.util.AttributeSet;

public class LinedEditText extends androidx.appcompat.widget.AppCompatEditText {
    private Rect rect;
    private Paint paint;
    private Paint bgPaint;
    private int gutterWidth;

    public LinedEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        rect = new Rect();
        paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setStyle(Paint.Style.FILL);
        paint.setColor(Color.parseColor("#9E9E9E"));
        paint.setTextSize(getTextSize() * 0.9f);
        
        bgPaint = new Paint();
        bgPaint.setColor(Color.parseColor("#EEEEEE"));

        // Set padding to make room for line numbers
        gutterWidth = (int) (paint.measureText("9999") + 20);
        setPadding(gutterWidth + 10, getPaddingTop(), getPaddingRight(), getPaddingBottom());
    }

    @Override
    protected void onDraw(Canvas canvas) {
        // Draw gutter background
        int scrollY = getScrollY();
        canvas.drawRect(0, scrollY, gutterWidth, scrollY + getHeight(), bgPaint);

        int baseline = getBaseline();
        int lineCount = getLineCount();
        
        // Optimize drawing to only visible lines
        int firstVisibleLine = getLayout().getLineForVertical(scrollY);
        int lastVisibleLine = getLayout().getLineForVertical(scrollY + getHeight());

        for (int i = firstVisibleLine; i <= lastVisibleLine; i++) {
            int lineNum = i + 1;
            String lineNumStr = String.valueOf(lineNum);
            
            // Calculate Y position for the baseline of the current line
            int yPos = getLayout().getLineBaseline(i);
            
            canvas.drawText(lineNumStr, gutterWidth - paint.measureText(lineNumStr) - 10, yPos, paint);
        }
        super.onDraw(canvas);
    }
}

File: app/src/main/java/bin/mg/main/ui/editor/LinedEditText.kt
package bin.mg.main.ui.editor

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatEditText

class LinedEditText @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null
) : AppCompatEditText(context, attrs) {

    private val rect = android.graphics.Rect()
    private val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.FILL
        color = Color.parseColor("#9E9E9E")
        textSize = this@LinedEditText.textSize * 0.9f
    }
    private val bgPaint = Paint().apply {
        color = Color.parseColor("#EEEEEE")
    }
    private val gutterWidth: Int

    init {
        gutterWidth = (paint.measureText("9999") + 20).toInt()
        setPadding(gutterWidth + 10, paddingTop, paddingRight, paddingBottom)
    }

    override fun onDraw(canvas: Canvas) {
        val scrollY = scrollY
        canvas.drawRect(0f, scrollY.toFloat(), gutterWidth.toFloat(), (scrollY + height).toFloat(), bgPaint)

        val layout = layout ?: return
        val firstVisibleLine = layout.getLineForVertical(scrollY)
        val lastVisibleLine = layout.getLineForVertical(scrollY + height)

        for (i in firstVisibleLine..lastVisibleLine) {
            val lineNum = (i + 1).toString()
            val yPos = layout.getLineBaseline(i)
            canvas.drawText(lineNum, gutterWidth - paint.measureText(lineNum) - 10, yPos.toFloat(), paint)
        }
        super.onDraw(canvas)
    }
}
